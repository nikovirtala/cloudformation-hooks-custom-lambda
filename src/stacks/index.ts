import { join } from "node:path";
import { Stack, type StackProps, aws_cloudformation, aws_iam, aws_lambda, aws_lambda_nodejs } from "aws-cdk-lib";
import type { Construct } from "constructs";

export class MyStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const c = new aws_lambda_nodejs.NodejsFunction(this, "ChangeSetFunction", {
            applicationLogLevelV2: aws_lambda.ApplicationLogLevel.DEBUG,
            architecture: aws_lambda.Architecture.ARM_64,
            bundling: {
                format: aws_lambda_nodejs.OutputFormat.ESM,
                target: "esnext",
            },
            entry: join(import.meta.dirname, "../handlers/change-set.ts"),
            loggingFormat: aws_lambda.LoggingFormat.JSON,
            runtime: aws_lambda.Runtime.NODEJS_22_X,
        });

        const t = new aws_lambda_nodejs.NodejsFunction(this, "TemplateFunction", {
            applicationLogLevelV2: aws_lambda.ApplicationLogLevel.DEBUG,
            architecture: aws_lambda.Architecture.ARM_64,
            bundling: {
                format: aws_lambda_nodejs.OutputFormat.ESM,
                target: "esnext",
            },
            entry: join(import.meta.dirname, "../handlers/template.ts"),
            loggingFormat: aws_lambda.LoggingFormat.JSON,
            runtime: aws_lambda.Runtime.NODEJS_22_X,
        });

        const hooksRole = new aws_iam.Role(this, "HooksRole", {
            assumedBy: new aws_iam.ServicePrincipal("hooks.cloudformation.amazonaws.com"),
            path: "/",
            inlinePolicies: {
                LambdaInvokerHookPolicy: new aws_iam.PolicyDocument({
                    statements: [
                        new aws_iam.PolicyStatement({
                            effect: aws_iam.Effect.ALLOW,
                            actions: ["lambda:InvokeFunction"],
                            resources: [c.functionArn, t.functionArn],
                        }),
                    ],
                }),
            },
        });

        new aws_cloudformation.CfnLambdaHook(this, "ChangeSetHook", {
            alias: "Test::Lambda::ChangeSetHook",
            executionRole: hooksRole.roleArn,
            failureMode: "WARN",
            hookStatus: "ENABLED",
            lambdaFunction: c.functionArn,
            stackFilters: {
                filteringCriteria: "ALL",
                stackNames: {
                    exclude: [this.stackName],
                },
            },
            targetOperations: ["CHANGE_SET"],
        });

        new aws_cloudformation.CfnLambdaHook(this, "StackHook", {
            alias: "Test::Lambda::StackHook",
            executionRole: hooksRole.roleArn,
            failureMode: "WARN",
            hookStatus: "ENABLED",
            lambdaFunction: t.functionArn,
            stackFilters: {
                filteringCriteria: "ALL",
                stackNames: {
                    exclude: [this.stackName],
                },
            },
            targetOperations: ["STACK"],
        });
    }
}
