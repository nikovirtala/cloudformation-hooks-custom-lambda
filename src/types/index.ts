import { Handler } from "aws-lambda";

export interface CloudFormationHookRequest {
    awsAccountId: string;
    stackId: string;
    changeSetId?: string;
    hookTypeName: string;
    hookTypeVersion: string;
    hookModel: {
        LambdaFunction: string;
    };
    actionInvocationPoint: "CREATE_PRE_PROVISION" | "UPDATE_PRE_PROVISION" | "DELETE_PRE_PROVISION";
    requestData: {
        targetName: string;
        targetType: string;
        targetLogicalId: string;
        targetModel?: {
            resourceProperties: Record<string, any>;
            previousResourceProperties?: Record<string, any>;
        };
        payload?: string; // S3 Presigned URL for stack and change set operations
    };
    requestContext: {
        invocation: number;
        callbackContext?: any;
    };
    clientRequestToken: string;
}

export interface CloudFormationHookResponse {
    hookStatus: "SUCCESS" | "FAILED" | "IN_PROGRESS";
    errorCode?: "NonCompliant" | "InternalFailure";
    message: string;
    clientRequestToken: string;
    callbackContext?: any;
    callbackDelaySeconds?: number;
}

export type CloudFormationHookHandler = Handler<CloudFormationHookRequest, CloudFormationHookResponse>;
