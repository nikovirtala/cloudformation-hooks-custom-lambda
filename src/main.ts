import { MyStack } from "@/stacks";
import { App } from "aws-cdk-lib";

const app = new App();

new MyStack(app, "cloudformation-hooks-custom-lambda");

app.synth();
