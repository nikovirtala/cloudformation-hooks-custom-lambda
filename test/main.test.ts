import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { expect, test } from "vitest";
import { MyStack } from "../src/stacks";

test("Snapshot", () => {
    const app = new App();
    const stack = new MyStack(app, "test");

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
});
