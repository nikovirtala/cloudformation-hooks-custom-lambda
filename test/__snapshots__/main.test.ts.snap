// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ChangeSetFunction2CB6D94D": {
      "DependsOn": [
        "ChangeSetFunctionServiceRole87D5FB3A",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "9b8cff59c6c4bbbdf5d3d7475a58cf88234a285f215f5a777291ee241fcbd6b8.zip",
        },
        "Handler": "index.handler",
        "LoggingConfig": {
          "ApplicationLogLevel": "DEBUG",
          "LogFormat": "JSON",
        },
        "Role": {
          "Fn::GetAtt": [
            "ChangeSetFunctionServiceRole87D5FB3A",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ChangeSetFunctionServiceRole87D5FB3A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ChangeSetHook": {
      "Properties": {
        "Alias": "Test::Lambda::ChangeSetHook",
        "ExecutionRole": {
          "Fn::GetAtt": [
            "HooksRole0184100C",
            "Arn",
          ],
        },
        "FailureMode": "WARN",
        "HookStatus": "ENABLED",
        "LambdaFunction": {
          "Fn::GetAtt": [
            "ChangeSetFunction2CB6D94D",
            "Arn",
          ],
        },
        "StackFilters": {
          "FilteringCriteria": "ALL",
          "StackNames": {
            "Exclude": [
              "test",
            ],
          },
        },
        "TargetOperations": [
          "CHANGE_SET",
        ],
      },
      "Type": "AWS::CloudFormation::LambdaHook",
    },
    "HooksRole0184100C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "hooks.cloudformation.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "lambda:InvokeFunction",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::GetAtt": [
                        "ChangeSetFunction2CB6D94D",
                        "Arn",
                      ],
                    },
                    {
                      "Fn::GetAtt": [
                        "TemplateFunction8B010058",
                        "Arn",
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaInvokerHookPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StackHook": {
      "Properties": {
        "Alias": "Test::Lambda::StackHook",
        "ExecutionRole": {
          "Fn::GetAtt": [
            "HooksRole0184100C",
            "Arn",
          ],
        },
        "FailureMode": "WARN",
        "HookStatus": "ENABLED",
        "LambdaFunction": {
          "Fn::GetAtt": [
            "TemplateFunction8B010058",
            "Arn",
          ],
        },
        "StackFilters": {
          "FilteringCriteria": "ALL",
          "StackNames": {
            "Exclude": [
              "test",
            ],
          },
        },
        "TargetOperations": [
          "STACK",
        ],
      },
      "Type": "AWS::CloudFormation::LambdaHook",
    },
    "TemplateFunction8B010058": {
      "DependsOn": [
        "TemplateFunctionServiceRoleD527DE0A",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c7a3fdc3f61195793b5c5e5c28ce0def8e67b08144154d0f29b1a00513739f9e.zip",
        },
        "Handler": "index.handler",
        "LoggingConfig": {
          "ApplicationLogLevel": "DEBUG",
          "LogFormat": "JSON",
        },
        "Role": {
          "Fn::GetAtt": [
            "TemplateFunctionServiceRoleD527DE0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TemplateFunctionServiceRoleD527DE0A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
